### ISSUE TITLE
Enable parameterized Graph IR nodes for gates that require runtime parameters

### ISSUE THREAD

**Issue by amccaskey** [ORNL-QCI/xacc#13]
We need to update our QuantumCircuit Graph to enable nodes that can be parameterized. 

For example - a kernel that takes a double representing an angle for a rotation gate
```cpp
__qpu__ void kernel(double phi) {
   qbit qreg[1];
   Rz(qreg[0], phi);
}
```
I think we can do this with extra CircuitNode parameters, then Accelerators executing GraphIR can be modified to search for these runtime parameters in the graph and update them with the correct values provided at runtime.

**Comment by amccaskey** [comment]
Reference #5

**Comment by amccaskey** [comment]
One problem - double is given to us as a string... How to store a placeholder node when we don't have it as an actual type?

**Comment by amccaskey** [comment]
We can use decltype from the variadic template arguments passed to getKernel.

At compile time, we can just keep a reference to the type string and variable name, and setup up a new CircuitNode that is parameterized.

**Comment by amccaskey** [comment]
Just had an idea...

The GraphIR CircuitNodes can keep track of parameter variable names. Ie above, the Rz node needs to have reference to the variable name 'phi'. 

Then at the API level, users construct Kernel's with variadic list of kernel function argument types, with the function itself taking as argument the name of the kernel, and the list of variable string names (or we can automate this, actually I like that better (stream of consciousness here))

Then the kernel lambda has reference to the types it takes, and we write the lamda to fill a mapping of kernel argument string names to actual values, and the accelerator can query this map when it hits a gate that is parameterized.

**Comment by amccaskey** [comment]
The above would look something like this: 

```cpp
const std::string src = "__qpu__ void kernel(double phi) {\n"
   "   qbit qreg[1];\n"
   "   Rz(qreg[0], phi);\n"
   "}\n";
```
then in main we'd have
```cpp
auto k = program.getKernel<double>("kernel");
k(3.14);
```

**Comment by amccaskey** [comment]
addressed as of 617bbecef863042c875086a4958443a4fbfc6569

leaving open for now in case of bugs
